### PR TITLE
refactor!: Refactor to make easy improvements

### PR DESCRIPTION
**Pull request by taorepoara**
Has explained in the next comment, this refactor makes this project simpler to use and to improve: https://github.com/yanganto/struct-patch/issues/24#issuecomment-2254496376

I've based my work on the serde way of parsing AST into structs to make it more usable.

I've also changed the way I manage attributes to make it easier to add new ones.

### CODE REVIEW

File: struct-patch-derive/src/lib.rs [hunk 1]
```diff
@@ -1,294 +1,421 @@
...
+use proc_macro2::{Ident, Span, TokenStream};
+use quote::{quote, ToTokens};
+use syn::{
+    meta::ParseNestedMeta, parenthesized, spanned::Spanned, DeriveInput, Error, LitStr, Type,
+};
+
+const PATCH: &str = "patch";
+const NAME: &str = "name";
+const ATTRIBUTE: &str = "attribute";
+const TYPE: &str = "type";
```
Comment (yanganto) [inline struct-patch-derive/src/lib.rs:13]: The `TYPE` do the same thing as `NAME`, we used `patch_name` before.  It is good to drop `TYPE` and keep only `NAME` here.

File: struct-patch-derive/src/lib.rs [hunk 2]
```diff
@@ -1,294 +1,421 @@
...
-                #patch_struct_name {
-                    #(
-                        #field_names: None,
-                    )*
+            attr.parse_nested_meta(|meta| {
+                let path = meta.path.to_string();
+                match path.as_str() {
+                    SKIP => {
+                        // #[patch(skip)]
+                        skip = true;
```
Comment (yanganto) [inline struct-patch-derive/src/lib.rs:303]: Could we return `Ok(None)` here? Such that we can remove the `skip` variable.

**Comment by yanganto** [comment]
Could we remove `attribute`?  and then the syntax will be simple
```
#[patch(attribute(derive(Deserialize, Debug), serde(rename = "foo"))]
```
->
```
#[patch(derive(Deserialize, Debug), serde(rename = "foo")]
```

**Review by yanganto — CHANGES_REQUESTED** [review]
Could we remove `attribute`?  and then the syntax will be simple
```
#[patch(attribute(derive(Deserialize, Debug), serde(rename = "foo"))]
```
->
```
#[patch(derive(Deserialize, Debug), serde(rename = "foo")]
```

File: struct-patch-derive/src/lib.rs (hunk 1 above)
Comment (taorepoara) [reply struct-patch-derive/src/lib.rs:13]: I changed name to type here to do not confuse between the name attribute on the struc that defines the name of the patch struct and the one on the field that define the type of field in the patch struct.

Let me know what you prefer

File: struct-patch-derive/src/lib.rs (hunk 2 above)
Comment (taorepoara) [reply struct-patch-derive/src/lib.rs:303]: I tried but we can't since we are in the `attr.parse_nested_meta` callback...

**Comment by taorepoara** [comment]
> Could we remove `attribute`? and then the syntax will be simple
> 
> ```
> #[patch(attribute(derive(Deserialize, Debug), serde(rename = "foo"))]
> ```
> 
> ->
> 
> ```
> #[patch(derive(Deserialize, Debug), serde(rename = "foo")]
> ```

That's true it would be so much simple !

We just have to be careful to avoid the possibility between the attributes and the patch configuration attributes, but if there is no way to have conflict I can do that.

I also thought about this syntax (not sure if it's possible).

What do you think?

```
#[patch(
  #[derive(Deserialize, Debug)],
  #[serde(rename = "foo")]
)]
```

File: struct-patch-derive/src/lib.rs (hunk 1 above)
Comment (yanganto) [reply struct-patch-derive/src/lib.rs:13]: How about type_rename, I think one is good we can have a good parameter name

Comment (taorepoara) [reply struct-patch-derive/src/lib.rs:13]: I was not found of the type highlight two 😅

I think that one word is better, so I'll keep name like you suggested in first comment 😉

**Comment by yanganto** [comment]
> ```
> #[patch(
>   #[derive(Deserialize, Debug)]
>   #[serde(rename = "foo")]
> )]
> ```
It is good and without `,`, just parse everything between `#[` and `]` as inputs.
Such that if the developer forgot to put `,`, it will not mater.

**Comment by taorepoara** [comment]
> > ```
> > #[patch(
> >   #[derive(Deserialize, Debug)]
> >   #[serde(rename = "foo")]
> > )]
> > ```
> 
> It is good and without `,`, just parse everything between `#[` and `]` as inputs. Such that if the developer forgot to put `,`, it will not mater.

Yes ok, I'll try that. This could allow code injection but it's his own code and I'll see how the formatter manage that

File: struct-patch-derive/src/lib.rs (hunk 1 above)
Comment (yanganto) [reply struct-patch-derive/src/lib.rs:13]: By the way, Github will do hightlight for the `type`  in macro
```rust
#[patch]
struct A {
  #[patch(type=BB)]
  field_b: B,
  #[patch(ignore)]
  field_c: C,
}
```

**Comment by taorepoara** [comment]
I could not implement a good solution for the attribute since the [`parse_nested_meta`](https://github.com/taorepoara/struct-patch/blob/cff660f645611e99669d9df5cf358c5c74318f59/struct-patch-derive/src/lib.rs#L194) try to parse the sub attributes names to Iden which fails on `#` character.

This could be managed with more low level parsing but could not made it this morning.

I can look at that but only in a few days. I let the `attribute` for now.

**Review by yanganto — APPROVED** [review]

**Comment by yanganto** [comment]
Hi @taorepoara, looks good to me.
Is there anything you want to refactor in the PR?

**Comment by taorepoara** [comment]
> Hi @taorepoara, it looks good to me. Is there still anything you want to refactor in the PR?

No, that all for this one. I'm working on another PR for other features and I'll try to implement the `#[patch(#[my-attribute])]` syntax later in another PR.

**Pull request merged**